include/sys/_pmap.h

```diff
@@ -0,0 +1,108 @@
+#ifndef _SYS__PMAP_H_
+#define _SYS__PMAP_H_
+
+#ifndef _MACHDEP
+#error "Do not use this header file outside kernel machine dependent code!"
+#endif
+
+#include <sys/mimiker.h>
+#include <sys/mutex.h>
+#include <sys/vm.h>
+#include <machine/pmap.h>
+
+struct pmap {
+  mtx_t mtx;                      /* protects all fields in this structure */
+  asid_t asid;                    /* address space identifier */
+  paddr_t pde;                    /* directory page table physical address */
+  vm_pagelist_t pte_pages;        /* pages we allocate in page table */
+  TAILQ_HEAD(, pv_entry) pv_list; /* all pages mapped by this physical map */
+
+  /* Machine-dependent part */
+  pmap_md_t md;
+};
+
+typedef struct pv_entry {
+  TAILQ_ENTRY(pv_entry) pmap_link; /* link on `pmap::pv_list` */
+  TAILQ_ENTRY(pv_entry) page_link; /* link on `vm_page::pv_list` */
+  pmap_t *pmap;                    /* page is mapped in this pmap */
+  vaddr_t va;                      /* under this address */
+} pv_entry_t;
+
+/*
+ * Machine dependent flags passed to `pmap_kenter` and `pmap_enter`.
+ */
+
+#define _PMAP_KERNEL (1 << 24)
+
+/*
+ * Each target must implement the following interface
+ * along with the `pmap_md_t` struct and a handful of macros:
+ *
+ *  - `PAGE_TABLE_DEPTH`: depth of the address translation structure
+ *  - `DMAP_BASE`: virtual address of the beginning of DMAP
+ *  - `MAX_ASID`: maximum possible ASID
+ *  - `PTE_KERNEL_EMPTY`: PTE value used for kernel empty page table entries
+ *  - `PTE_USER_EMPTY`: PTE value used for user empty page table entries
+ *  - `PTE_SET_ON_REFERENCED`: PTE bits to set while marking a page
+ *    as referenced
+ *  - `PTE_CLR_ON_REFERENCED`: PTE bits to clear while marking a page
+ *    as referenced
+ *  - `PTE_SET_ON_MODIFIED`: PTE bits to set while marking a page as modified
+ *  - `PTE_CLR_ON_MODIFIED`: PTE bits to clear while marking a page as modified
+ *  - `GROWKERNEL_STRIDE`: stride used while expanding the kernel virtual
+ *    address space
+ *
+ * Besides, before the pmap module can be used, each target must:
+ *
+ *  - build DMAP and set the `dmap_paddr_base` and `dmap_paddr_end`
+ *  - set the base address of the kernel page directory (`kernel_pde`)
+ *
+ * The pmap module requires the following machine-dependent types to be defined:
+ *
+ *  - `pte_t`: page table entry
+ *  - `pde_t`: page directory entry
+ *  - `asid_t`: address space identifier
+ */
+
+extern paddr_t dmap_paddr_base;
+extern paddr_t dmap_paddr_end;
+
+/*
+ * Page directory.
+ */
+
+static inline bool pde_valid_p(pde_t *pdep);
+pde_t pde_make(int lvl, paddr_t pa);
+static inline pde_t *pde_ptr(paddr_t pd, int lvl, vaddr_t va);
+
+/*
+ * Some architectures may require broadcasting changes in the top page directory
+ * of the kernel. This function is called whenever such a change occurs.
+ */
+void broadcast_kernel_top_pde(vaddr_t va, pde_t pde);
```

Comment:
We have some convention(s) for function names. I think that this one should be prefixed by `pmap_`.
@cahirwpz WDYT?

Reply:
@cahirwpz suggested removing this function from this PR. We are still unsure about this function and prefer to postpone the final decision to #1256.